apisix/plugins/lago.lua

```diff
@@ -0,0 +1,224 @@
+
+-- According to https://getlago.com/docs/api-reference/events/batch, the maximum batch size is 100,
+-- so we have to override the default batch size to make it work out of the box，the plugin does
+-- not set a maximum limit, so if Lago relaxes the limit, then user can modify it
+-- to a larger batch size
+-- This does not affect other plugins, schema is appended after deep copy
+schema.properties.batch_max_size.default = 100
+
+
+local _M = {
+    version = 0.1,
+    priority = 415,
+    name = plugin_name,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf, schema_type)
+    local check = {"endpoint_addrs"}
+    core.utils.check_https(check, conf, plugin_name)
+    core.utils.check_tls_bool({"ssl_verify"}, conf, plugin_name)
+
+    return core.schema.check(schema, conf)
+end
+
+
+local function send_http_data(conf, data)
+    local params = {
+        headers = {
+            ["Content-Type"] = "application/json",
+            ["Authorization"] = "Bearer " .. conf.token,
+        },
+        keepalive = conf.keepalive,
+        ssl_verify = conf.ssl_verify,
+        method = "POST",
+        body = core.json.encode(data)
+    }
+
+    if conf.keepalive then
+        params.keepalive_timeout = conf.keepalive_timeout
+        params.keepalive_pool = conf.keepalive_pool
+    end
+
+    local httpc, err = http.new()
+    if not httpc then
+        return false, str_format("create http client error: %s", err)
+    end
+    httpc:set_timeout(conf.timeout)
+
+    -- select an random endpoint and build URL
+    local endpoint_url = conf.endpoint_addrs[math_random(#conf.endpoint_addrs)]..conf.endpoint_uri
+    local res, err = httpc:request_uri(endpoint_url, params)
+    if not res then
+        return false, err
+    end
+
+    if res.status >= 300 then
+        return false, str_format("lago api returned status: %d, body: %s",
+            res.status, res.body or "")
+    end
+
+    return true
+end
+
+
+function _M.log(conf, ctx)
+    -- build usage event
+    local event_transaction_id, err = core.utils.resolve_var(conf.event_transaction_id, ctx.var)
+    if err then
+        core.log.error("failed to resolve event_transaction_id, event dropped: ", err)
+        return
+    end
+
+    local event_subscription_id, err = core.utils.resolve_var(conf.event_subscription_id, ctx.var)
+    if err then
+        core.log.error("failed to resolve event_subscription_id, event dropped: ", err)
+        return
+    end
+
+    local entry = {
+        transaction_id = event_transaction_id,
+        external_subscription_id = event_subscription_id,
+        code = conf.event_code,
+        timestamp = ngx.req.start_time(),
+    }
+
+    if conf.event_properties and type(conf.event_properties) == "table" then
+        entry.properties = core.table.deepcopy(conf.event_properties)
+        for key, value in pairs(entry.properties) do
+            local new_val, err, n_resolved = core.utils.resolve_var(value, ctx.var)
```

Comment:
Does this `err` not require logging?

Reply:
When parsing fails or unsuccessfully replaces any variable, it doesn't replace the original value, so the user will see strings that retain the original expression, e.g. `xx ${xxx}`, so the error may be irrelevant?

This is the same code snippet used in the loki logger.